### PR TITLE
Location consistency

### LINKED ISSUE #404: Location Validation Framework
Provide a system-wide way to check that a given set of location ids (ie: StateID, DCode, TCode, etc...) is valid

### PR DESCRIPTION
**Pull request by koshalt**
Fixes #404 
Helps with #104 validation

### CODE REVIEW

File: kilkari/src/main/java/org/motechproject/nms/kilkari/service/impl/MctsBeneficiaryImportServiceImpl.java [hunk 1]
```diff
@@ -329,80 +316,45 @@ private Subscriber updateSubscriber(Subscriber subscriber, MctsBeneficiary benef
...
-                    }
-                }
-            }
-        }
+    private void setLocationFields(Map<String, Object> locations, MctsBeneficiary beneficiary) throws InvalidLocationException {
 
-        if (state == null || district == null) {
-            errors = "District and state must both be set for an MCTS beneficiary.";
+        if (locations.get(STATE) == null || locations.get(DISTRICT) == null) {
+            throw new InvalidLocationException("Missing mandatory state and district fields");
```
Comment (frankhuster) [inline kilkari/src/main/java/org/motechproject/nms/kilkari/service/impl/MctsBeneficiaryImportServiceImpl.java]: Make the logic better such that you know which field is missing.

File: region/src/main/java/org/motechproject/nms/region/service/LocationService.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+import org.motechproject.nms.region.domain.HealthSubFacility;
+import org.motechproject.nms.region.domain.State;
+import org.motechproject.nms.region.domain.Taluka;
+import org.motechproject.nms.region.domain.Village;
+import org.motechproject.nms.region.exception.InvalidLocationException;
+
+import java.util.Map;
+
+/**
+ * Location service to get and validation location data
```
Comment (frankhuster) [inline region/src/main/java/org/motechproject/nms/region/service/LocationService.java]: grammar?

File: region/src/main/java/org/motechproject/nms/region/service/impl/LocationServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+        this.healthFacilityService = healthFacilityService;
+        this.healthSubFacilityService = healthSubFacilityService;
+    }
+
+    @Override // NO CHECKSTYLE Cyclomatic Complexity
+    @SuppressWarnings("PMD")
+    public Map<String, Object> getLocations(Map<String, Object> locationMapping) throws InvalidLocationException {
+
+        Map<String, Object> locations = new HashMap<>();
+        do { // Note: this is not a true loop, just a better way to break and pass control
```
Comment (frankhuster) [inline region/src/main/java/org/motechproject/nms/region/service/impl/LocationServiceImpl.java]: kill loop and use `return locations` instead of `break`?

File: region/src/main/java/org/motechproject/nms/region/service/impl/VillageServiceImpl.java [hunk 4]
```diff
@@ -19,7 +19,7 @@
 
 
     @Override
-    public Village findByTalukaAndVcodeAndSvid(final Taluka taluka, final long vcode, final long svid) {
+    public Village findByTalukaAndVcodeAndSvid(final Taluka taluka, final Long vcode, final Long svid) {
```
Comment (frankhuster) [inline region/src/main/java/org/motechproject/nms/region/service/impl/VillageServiceImpl.java]: why the change to Long?

**Comment by frankhuster** [comment]
lgtm

File: kilkari/src/main/java/org/motechproject/nms/kilkari/service/impl/MctsBeneficiaryImportServiceImpl.java (hunk 1 above)
Comment (koshalt) [inline kilkari/src/main/java/org/motechproject/nms/kilkari/service/impl/MctsBeneficiaryImportServiceImpl.java]: fixed.

File: region/src/main/java/org/motechproject/nms/region/service/LocationService.java (hunk 2 above)
Comment (koshalt) [inline region/src/main/java/org/motechproject/nms/region/service/LocationService.java]: :flushed:

File: region/src/main/java/org/motechproject/nms/region/service/impl/LocationServiceImpl.java (hunk 3 above)
Comment (koshalt) [inline region/src/main/java/org/motechproject/nms/region/service/impl/LocationServiceImpl.java]: done
